google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/read_rows/StateMachine.java

```diff
@@ -0,0 +1,414 @@
+   *
+   * @return The last completed row.
+   * @throws IllegalStateException If the last chunk did not complete a row.
+   */
+  RowT consumeRow() {
+    Preconditions.checkState(currentState == AWAITING_ROW_CONSUME, "No row to consume");
+    RowT row = completeRow;
+    reset();
+    return row;
+  }
+
+  /** Checks if there is a complete to bew consumed. */
+  boolean hasCompleteRow() {
+    return currentState == AWAITING_ROW_CONSUME;
+  }
+  /**
+   * Checks if the state machine is in the middle of processing a row.
+   *
+   * @return True If there is a row in progress.
+   */
+  boolean isRowInProgress() {
+    return currentState != AWAITING_NEW_ROW;
+  }
+
+  private void reset() {
+    currentState = AWAITING_NEW_ROW;
+    rowKey = null;
+    familyName = null;
+    qualifier = null;
+    timestamp = 0;
+    labels = null;
+    expectedCellSize = 0;
+    remainingCellBytes = 0;
+    completeRow = null;
+
+    adapter.reset();
+  }
+
+  /**
+   * Base class for all of the state machine's internal states.
+   *
+   * <p>Each state can consume 2 events: lastScannedRowKey and a chunk. By default, the default
+   * implementation will just throw an IllegalStateException unless the subclass adds explicit
+   * handling for the event by overriding {@link #handleLastScannedRow(ByteString)} or {@link
+   * #handleChunk(CellChunk)}.
+   */
+  abstract static class State {
+    /**
+     * Accepts the last row key scanned by the server. And set its as the last complete row. This
+     * row should be treated specially because it doesn't actually contain data.
+     *
+     * @throws IllegalStateException If the subclass can't handle last scanned row events.
+     */
+    State handleLastScannedRow(ByteString rowKey) {
+      throw new IllegalStateException();
+    }
+
+    /**
+     * Accepts a new chunk and transitions to the next state.
+     *
+     * @param chunk The new chunk to process.
+     * @return The next state.
+     * @throws IllegalStateException If the subclass can't handle chunks.
+     * @throws InvalidInputException If the subclass determines that this chunk is invalid.
+     */
+    State handleChunk(CellChunk chunk) {
+      throw new IllegalStateException();
+    }
+  }
+
+  /**
+   * The default state when the state machine is awaiting a chunk to start a new row. It will notify
+   * the adapter of the new row and delegate to the AWAITING_NEW_CELL state to process the first
+   * Cell in the Chunk. Exit states: same as AWAITING_NEW_CELL
+   */
+  private final State AWAITING_NEW_ROW =
+      new State() {
+        @Override
+        State handleLastScannedRow(ByteString rowKey) {
+          completeRow = adapter.createScanMarkerRow(rowKey);
+          lastCompleteRowKey = rowKey;
+          return AWAITING_ROW_CONSUME;
+        }
+
+        @Override
+        State handleChunk(CellChunk chunk) {
+          validate(!chunk.getResetRow(), "AWAITING_NEW_ROW: can't reset");
```

Comment:
Thoughts on using a toString() method per State for this? Reduces the possibility of a copy paste error between states...

Reply:
Unfortunately the toString is useless here because these are anonymous classes